### PR TITLE
Adding implementation for EmpiricalDistribution

### PR DESCRIPTION
**Pull request by neerajprad**
Implements EmpiricalDistribution based on comments and suggestions from @fritzo and @eb8680.

The goal was the following:
 - Have a generic `EmpiricalDistribution` implementation that can accept importance weights.
 - Supporting `log_prob`. However, supporting this should not result in a performance penalty for cases where we do not need to compute `log_prob`. Hence any heavy pre-processing (like tensor comparisons) required when adding samples to support `log_prob` is bypassed.
 - Supporting fast `append` operation, without too much overhead - we append to a buffer which is lazily concatenated to an internal tensor when needed. 
 - Should be possible to re-use the same `EmpiricalDistribution` implementation. e.g. appending some value, computing the mean, appending some more values, re-computing the mean etc. There should be no assumption that the add operation precedes all statistical computation on the samples.

What is currently missing?
 - I have not figured out a way to support vectorized `log_prob`, i.e. the ability to score multiple samples at the same time. Doing so will likely require a different implementation that needs to pre-process samples as they are added, and will be slow. Currently, `log_prob` only accepts single values.

Addresses #942. 

Related: #943 - Alternate sketch that does not support `log_prob` or `importance_weights`.

### CODE REVIEW

File: pyro/distributions/empirical.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        if is_validation_enabled():
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = torch.Tensor if value.dtype in (torch.int32, torch.int64) else value.type()
+        # Apply default weight of 1.0.
+        if log_weight is None and weight is None:
+            log_weight = torch.tensor(1.0).type(weight_type).log()
```
Comment (martinjankowiak) [inline pyro/distributions/empirical.py]: why not just do zero directly without the expensive log?

Comment (neerajprad) [reply pyro/distributions/empirical.py]: :) Was thinking that it might be self documenting, but you are right. Will change.

File: pyro/distributions/util.py [hunk 2]
```diff
@@ -321,6 +321,20 @@ def matrix_triangular_solve_compat(b, A, upper=True):
         return b.trtrs(A, upper=upper)[0].view(b.size())
 
 
+def log_sum_exp(tensor, scale=1.0, dim=-1):
```
Comment (fritzo) [inline pyro/distributions/util.py]: The scale arg here seems a little weird. I'd at least move it to the right of `dim`, but consider removing it.

File: pyro/distributions/empirical.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    @property
+    def mean(self):
+        self._finalize()
+        if self._samples.dtype in (torch.int32, torch.int64):
+            raise ValueError("Mean for discrete empirical distribution undefined. Consider converting samples " +
+                             "to ``torch.float32`` or ``torch.float64``.")
+        weights = self._log_weights
+        for _ in range(self._samples.dim() - 1):
+            weights = weights.unsqueeze(-1)
+        return (log_sum_exp(weights, scale=self._samples, dim=0) - log_sum_exp(weights, dim=0)).exp()
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: It seems like it would be clearer to
```py
weights = (self._log_weights - log_sum_exp(self._log_weights)).exp()
weights = weights.reshape(weights.shape + (1,) * self.event_dim)
return (weights * self._samples).sum(0)
```
Then you could also remove the `scale` arg to `log_sum_exp()`

File: pyro/distributions/empirical.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        Number of samples that constitute the empirical distribution.
+
+        :return int: number of samples collected.
+        """
+        self._finalize()
+        if self._samples is None:
+            return 0
+        return self._samples.size(0)
+
+    def add(self, value, log_weight=None, weight=None):
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: nit: `weight` seems more standard than `log_weight`, so I'd be less surprised if if was to the left of `log_weight`

File: pyro/distributions/empirical.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        :param torch.Tensor log_weight: weight (optional) corresponding
+            to the sample.
+        """
+        if is_validation_enabled():
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = torch.Tensor if value.dtype in (torch.int32, torch.int64) else value.type()
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: This doesn't work with cuda int tensors. Consider using something that preserves CUDA and device, e.g.
```py
weight_type = value.new_empty(1).float().type() if ... else value.type()
```

File: pyro/distributions/empirical.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    def variance(self):
+        self._finalize()
+        if self._samples.dtype in (torch.int32, torch.int64):
+            raise ValueError("Variance for discrete empirical distribution undefined. Consider converting samples " +
+                             "to ``torch.float32`` or ``torch.float64``.")
+        deviation_squared = torch.pow(self._samples - self.mean, 2)
+        weights = self._log_weights
+        for _ in range(self._samples.dim() - 1):
+            weights = weights.unsqueeze(-1)
+        return (log_sum_exp(weights, scale=deviation_squared, dim=0) - log_sum_exp(weights, dim=0)).exp()
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: ditto: it seems like it would be clearer to
```py
weights = (self._log_weights - log_sum_exp(self._log_weights)).exp()
weights = weights.reshape(weights.shape + (1,) * self.event_dim)
mean = (weights * self._samples).sum(0)
return (weights * (self._samples - mean).pow(2)).sum(0)
```

File: pyro/distributions/empirical.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+@copy_docs_from(TorchDistribution)
+class Empirical(TorchDistribution):
+    r"""
+    Empirical distribution associated with the sampled data.
+    """
+
+    arg_constraints = {}
+
+    def __init__(self):
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: Can  you pass through the `validate_args` arg as other `TorchDistributions` do? Then you can use the idomatic `if self._validate_args` rather than `if is_validation_enabled()`?

File: pyro/distributions/empirical.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+from pyro.distributions.util import log_sum_exp
+
+
+@copy_docs_from(TorchDistribution)
+class Empirical(TorchDistribution):
+    r"""
+    Empirical distribution associated with the sampled data.
+    """
+
+    arg_constraints = {}
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: Is there a sensible `.support`? Does it make most sense to set `support = constraints.real`?

File: pyro/distributions/empirical.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+        # Seed the container tensors with the correct tensor types
+        if self._samples is None:
+            self._samples = value.new()
+            self._log_weights = log_weight.new()
+        # Append to the buffer list
+        self._samples_buffer.append(value)
+        self._weights_buffer.append(log_weight)
+
+    def sample(self, sample_shape=torch.Size()):
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: (low priority comment:)

Is it sensible to define `rsample()`? If samples are differentiable, then the results should be differentiable wrt samples. The result is not differentiable wrt `weights` though. @martinjankowiak wdyt?

File: pyro/distributions/empirical.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        selection_mask = self._samples.eq(value).contiguous().view(self.sample_size, -1)
+        # Return -Inf if value is outside the support.
+        if not selection_mask.any():
+            return self._log_weights.new_zeros(torch.Size()).log()
+        idxs = torch.arange(self.sample_size)[selection_mask.min(dim=-1)[0]]
+        log_probs = self._categorical.log_prob(idxs)
+        return log_sum_exp(log_probs)
+
+    @property
+    def batch_shape(self):
```
Comment (fritzo) [inline pyro/distributions/empirical.py]: I think you can omit this property this if you pass the `batch_shape` arg to `super(..., self).__init__()`

File: pyro/distributions/util.py (hunk 2 above)
Comment (neerajprad) [reply pyro/distributions/util.py]: The scale argument is adopted from scipy's [log_sum_exp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.logsumexp.html), though using it with tensorized scale was my idea. :) I suppose this will be more numerically stable than using the normalized weights and multiplying will have underflow issues - consider a low weight that is multiplied with a large scale factor.

We also have a `log_sum_exp` in `pyro.util` but using that leads to circular import issues.

Comment (fritzo) [reply pyro/distributions/util.py]: Ok. Can we move the `pyro.util` version to `pyro.distributions.util`?

File: pyro/distributions/empirical.py (hunk 3 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: See comment above. I think this will have underflow issues which can be avoided by including the scale factor. I can use the computation here directly rather than creating a generic utility, if you think a generic function is not warranted.

File: pyro/distributions/empirical.py (hunk 4 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: Will change.

File: pyro/distributions/empirical.py (hunk 7 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: Good idea, will change.

File: pyro/distributions/empirical.py (hunk 10 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: Will change.

File: pyro/distributions/empirical.py (hunk 8 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: Yeah, I suppose so.

File: pyro/distributions/empirical.py (hunk 5 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: Yes, let me change that.

File: pyro/distributions/empirical.py (hunk 6 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: I had it that way to begin with. Should I just move over the `log_sum_exp` logic here if the scale factor seems confusing?

File: pyro/distributions/util.py (hunk 2 above)
Comment (neerajprad) [reply pyro/distributions/util.py]: Sure. I think we can also remove that in a later cleanup if we start using this in `AbstractInfer`, since that is the only place it is used.

File: pyro/distributions/empirical.py (hunk 8 above)
Comment (eb8680) [reply pyro/distributions/empirical.py]: Are `enumerate_support` and `support` expected to satisfy any invariants in `TorchDistributions`?

Comment (fritzo) [reply pyro/distributions/empirical.py]: Some informal invariants are:
- `.enumerate_support()` should return a set of samples whose probabilities sum to 1
- arguments to `.log_prob()` should return `-inf` if outside of `.support`
- return values of `.sample()` and `.enumerate_support()` shoule lie in the `.support`

I don't know whether this is important; it just surprised me to see `.arg_constraints` without `.support`.

File: pyro/distributions/empirical.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/empirical.py]: I guess this is fine as is. It might be nice to avoid two different implementations of `log_sum_exp()` in the codebase, but we can always clean up in a follow-up PR.

File: pyro/distributions/empirical.py (hunk 6 above)
Comment (fritzo) [reply pyro/distributions/empirical.py]: It's up to you. I didn't realize the scale arg was provided by scipy.

File: pyro/distributions/empirical.py (hunk 3 above)
Comment (neerajprad) [reply pyro/distributions/empirical.py]: >  It might be nice to avoid two different implementations of log_sum_exp() in the codebase, but we can always clean up in a follow-up PR.

Already deleted, we are using the one from distributions now.

**Comment by neerajprad** [comment]
Will update with `.enumerate_support`.

**Comment by neerajprad** [comment]
@fritzo, @eb8680, @martinjankowiak - Thanks for reviewing this! There are a few things that I would like to change, and add another test, but that can wait for a subsequent PR. This should be good to merge otherwise.

File: pyro/distributions/empirical.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        :param torch.Tensor log_weight: weight (optional) corresponding
+            to the sample.
+        """
+        if self._validate_args:
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = value.new_empty(1).float().type() if value.dtype in (torch.int32, torch.int64) \
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/distributions/empirical.py]: good point, I don't know if there is a clean alternative other than lots of conditional logic.
